pkg/gateway/gateway.go

```diff
@@ -207,10 +227,101 @@ func New(ctx context.Context, cl zbus.Client, root string) (pkg.Gateway, error)
 		// this is not a failure because supporting of the gateway can happen
 		// later if the farmer set the correct network configuration!
 	}
-
+	go gw.nameContractsValidator()
 	return gw, nil
 }
 
+func (g *gatewayModule) getReservedDomain(domain string) (string, bool) {
+	g.domainLock.RLock()
+	defer g.domainLock.RUnlock()
+	v, ok := g.reservedDomains[domain]
+	return v, ok
+}
+
+func (g *gatewayModule) setReservedDomain(domain string, wlID string) {
+	g.domainLock.Lock()
+	defer g.domainLock.Unlock()
+	log.Debug().
+		Str("domain", domain).
+		Str("wlID", wlID).
+		Msg("setting domain")
+	g.reservedDomains[domain] = wlID
+}
+
+func (g *gatewayModule) deleteReservedDomain(domain string) {
+	g.domainLock.Lock()
+	defer g.domainLock.Unlock()
+	log.Debug().
+		Str("domain", domain).
+		Msg("deleting domain")
+	delete(g.reservedDomains, domain)
+}
+
+func (g *gatewayModule) copyReservedDomain() map[string]string {
+	g.domainLock.RLock()
+	defer g.domainLock.RUnlock()
+	res := make(map[string]string, len(g.reservedDomains))
+	for k, v := range g.reservedDomains {
+		res[k] = v
+	}
+	return res
+}
+
+func (g *gatewayModule) validateNameContracts() error {
+	ctx, cancel := context.WithTimeout(context.Background(), validationPeriod/2)
+	defer cancel()
+	e := stubs.NewProvisionStub(g.cl)
+	networker := stubs.NewNetworkerStub(g.cl)
+	cfg, err := networker.GetPublicConfig(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "gateway is not supported on this node, why are names are getting validated?")
+	}
+
+	baseDomain := cfg.Domain
+	if baseDomain == "" {
+		return errors.New("why is domain is empty while validating name contracts?")
+	}
+	reservedDomains := g.copyReservedDomain()
+	for domain, id := range reservedDomains {
+		wlID := gridtypes.WorkloadID(id)
+		twinID, _, _, err := wlID.Parts()
+		if err != nil {
+			log.Error().
+				Err(err).
+				Msgf("failed to parse wlID %s parts", id)
+			continue
+		}
+		if !strings.HasSuffix(domain, baseDomain) {
+			// a fqdn workload, skip validating it
+			continue
+		}
+		name := strings.TrimSuffix(domain, fmt.Sprintf(".%s", baseDomain))
+		if err := g.validateNameContract(name, twinID); err != nil {
```

Comment:
You try to decomission a name reservation on any error. IMHO validateNameContract should return specific set of errors (define some Err<Something>) where u can check. the decomission should happen always if the name contract does not exist or have been canceled. but not when you have a connection issue for example